govcd/nsxt_nat_rule.go

```diff
@@ -0,0 +1,289 @@
+		return err
+	}
+
+	apiVersion = elevateNsxtNatRuleApiVersion(apiVersion, client)
+
+	if nsxtNat.NsxtNatRule.ID == "" {
+		return fmt.Errorf("cannot delete NSX-T NAT rule without ID")
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, nsxtNat.edgeGatewayId), nsxtNat.NsxtNatRule.ID)
+	if err != nil {
+		return err
+	}
+
+	err = client.OpenApiDeleteItem(apiVersion, urlRef, nil, nil)
+
+	if err != nil {
+		return fmt.Errorf("error deleting NSX-T NAT Rule: %s", err)
+	}
+
+	return nil
+}
+
+// IsEqualTo allows to check if a rule has exactly the same fields (except ID) to the supplied rule
+// This validation is very tricky because minor version changes impact how fields are return.
+// This function relies on most common and stable fields:
+// * Name
+// * Enabled
+// * Description
+// * ExternalAddresses
+// * InternalAddresses
+// * ApplicationPortProfile.ID
+func (nsxtNat *NsxtNatRule) IsEqualTo(rule *types.NsxtNatRule) bool {
+	return natRulesEqual(nsxtNat.NsxtNatRule, rule)
+}
+
+// natRulesEqual is a helper to check if first and second supplied rules are exactly the same (except ID)
+func natRulesEqual(first, second *types.NsxtNatRule) bool {
+	util.Logger.Println("comparing NAT rule:")
+	util.Logger.Printf("%+v\n", first)
+	util.Logger.Println("against:")
+	util.Logger.Printf("%+v\n", second)
+
+	// Being an org user always returns logging as false - therefore cannot compare it.
+	// first.Logging == second.Logging &&
+
+	// These fields are returned or not returned depending on version and it is impossible to be 100% sure a minor
+	// patch does not break such comparison
+	// first.DnatExternalPort == second.DnatExternalPort &&
+	// first.SnatDestinationAddresses == second.SnatDestinationAddresses &&
+	//
+	if first.Name == second.Name &&
+		first.Enabled == second.Enabled &&
+		first.Description == second.Description &&
+
+		first.ExternalAddresses == second.ExternalAddresses &&
+		first.InternalAddresses == second.InternalAddresses &&
+		// Match either both application profiles being empty/nil, or both having the same value
+		((first.ApplicationPortProfile == second.ApplicationPortProfile) ||
+			(first.ApplicationPortProfile != nil && second.ApplicationPortProfile != nil && first.ApplicationPortProfile.ID == second.ApplicationPortProfile.ID)) {
+
+		return true
+	}
+
+	return false
+}
+
+// elevateNsxtNatRuleApiVersion helps to elevate API version to consume newer NSX-T NAT Rule features
+// API V35.2+ support new fields FirewallMatch and Priority
+// API V36.0+ supports new RuleType - REFLEXIVE
+func elevateNsxtNatRuleApiVersion(apiVersion string, client *Client) string {
+
+	// Fields FirewallMatch and Priority require API version 35.2 to be set therefore version is elevated if API supports
+	if client.APIVCDMaxVersionIs(">= 35.2") {
+		apiVersion = "35.2"
+	}
+
+	// RuleType REFLEXIVE requires API V36.0
+	if client.APIVCDMaxVersionIs(">= 36.0") {
+		apiVersion = "36.0"
+	}
+
+	return apiVersion
```

Comment:
This function assumes that `apiVersion` contains the valid version to be used when the client version is < 35.2.
But if `apiVersion` is empty or invalid, we will have an unrecoverable error.
Possibly we should retrieve apiVersion within `elevateNsxtNatRuleApiVersion` and return `(string error)`. 
Or you may have a better idea.

Reply:
The point with this function was that I would always feed it with output of 
```
endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointNsxtNatRules
apiVersion, err := client.checkOpenApiEndpointCompatibility(endpoint)
```

However. I had in mind that one day - if such approach of "stepped" elevation (elevation to known verified API versions when possible) could be more generic for those resources that need it. 

I have made a second attempt and introduced a function `client.getOpenApiHighestElevatedVersion` for internal use which is similar to `checkOpenApiEndpointCompatibility`, but is capable of picking highest possible version from a predefined list `endpointElevatedApiVersions`. 

We could just use "highest" version supported by VCD API, but that is risky as pointing it to future VCD version might just break it (if some field is removed in future API version).

Have a look and share if it looks better. This is yet hard to comprehend if it is a silver bullet as there is just one use case, but this was my early vision of how it could look in a more generic fashion.